### PR TITLE
PyTorch 1.11.0 and Python 3.10 and lief dev

### PR DESCRIPTION
**Pull request by hmaarrfk**
Closes: https://github.com/conda-forge/pytorch-cpu-feedstock/pull/72

Checklist
* [ ] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [ ] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [ ] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [ ] Ensured the license file is being packaged.

### CODE REVIEW

**Comment by hmaarrfk** [comment]
@rgommers I'm not sure if you've seen this error on OSX before.

**Comment by rgommers** [comment]
It doesn't look familiar. If I read correctly, `-Werror` is turned on for some third-party submodules, and it's complaining about a `--sysroot`  flag?
```
clang-12: error: argument unused during compilation: '--sysroot=/Applications/Xcode_12.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.9.sdk' [-Werror,-Wunused-command-line-argument]
```

**Comment by hmaarrfk** [comment]
Thank you. I forget why we did add the syroot. Let me see if removing it helps. I hope it doesn't hurt linux.

**Comment by h-vetinari** [comment]
Pretty cool that all the non-CUDA builds seem to run through with time to spare - should allow building them in CI consistently. Great work overall; hope the ctng-PR can get merged soon(-ish).

PS. Build_num should go back to 0 here.

**Comment by hmaarrfk** [comment]
@h-vetinari let me know if you approve of these changes. Hopefully the locally run GPU builds should finish in the coming days.

File: recipe/conda_build_config.yaml [hunk 1]
```diff
@@ -1,10 +1,19 @@
...
-  - 7                # [linux64]
-cxx_compiler_version:  # [linux64]
-  - 7                # [linux64]
-c_compiler:    # [win]
-  - vs2019     # [win]
-cxx_compiler:  # [win]
-  - vs2019     # [win]
+
+# Need to downgrade compiler for nccl until it gets patched on conda forge
+# https://github.com/NVIDIA/nccl/issues/494#issuecomment-881074063
```
Comment (h-vetinari) [inline recipe/conda_build_config.yaml]: Is that downgrade still necessary now that you picked the respective commit into the ctng-compiler feedstock?

Comment (h-vetinari) [reply recipe/conda_build_config.yaml]: nvm, sorry, was reviewing commit by commit 🙈

Comment (hmaarrfk) [reply recipe/conda_build_config.yaml]: Where are you seeing this? You might have an out of date view. I forced push and cleand it up
```
$ cat conda_build_config.yaml
blas_impl:
  - mkl                        # [x86 or x86_64]
```

File: recipe/meta.yaml [hunk 2]
```diff
@@ -99,19 +99,9 @@ outputs:
...
         - python
         - {{ pin_compatible('numpy') }}
-        # hmaarrfk 2021/12/28
-        # it seems that pytorch uses some aspects of setuptools
-        # that were removed in version 59.6
-        # I'm not really inclined to patch this out since
-        # this is a pretty easy build time dependency
-        # https://github.com/pytorch/pytorch/issues/69894
-        - setuptools <59.6
+        - setuptools
```
Comment (h-vetinari) [inline recipe/meta.yaml:102]: Is pytorch still run-depending on setuptools at all now?

Comment (hmaarrfk) [reply recipe/meta.yaml:102]: tbh, I'm really not sure.

Comment (hmaarrfk) [reply recipe/meta.yaml:102]: I remember i did try to audit it:
A grep command finds
```~/git/pytorch [v1.11.0|✔]
17:38 $ grep "import setuptools\|from setuptools" . -R
./tools/generate_torch_version.py:from setuptools import distutils  # type: ignore[import]
./tools/build_pytorch_libs.py:from setuptools import distutils  # type: ignore[import]
./tools/coverage_plugins_package/setup.py:import setuptools  # type: ignore[import]
./benchmarks/operator_benchmark/pt_extension/setup.py:from setuptools import setup
./torch/utils/cpp_extension.py:import setuptools
./torch/utils/cpp_extension.py:from setuptools.command.build_ext import build_ext
./torch/utils/cpp_extension.py:        >>> from setuptools import setup
./torch/utils/cpp_extension.py:        >>> from setuptools import setup
./torch/utils/cpp_extension.py:        from setuptools import distutils
./test/test_spectral_ops.py:from setuptools import distutils
./test/cpp_extensions/no_python_abi_suffix_test/setup.py:from setuptools import setup
./test/cpp_extensions/setup.py:from setuptools import setup
./setup.py:from setuptools import setup, Extension, find_packages
./setup.py:from setuptools.dist import Distribution
./setup.py:import setuptools.command.build_ext
./setup.py:import setuptools.command.install
./setup.py:import setuptools.command.sdist
```

But now, do you really want to ensure that cpp_extensions isn't imported?

Comment (h-vetinari) [reply recipe/meta.yaml:102]: My understanding from @rgommers in #80 was that the runtime-dependency should have been removed?

Comment (hmaarrfk) [reply recipe/meta.yaml:102]: i understood that unpinning was fine in 1.11.

i did not underatand that he meant to remove it outright

Comment (h-vetinari) [reply recipe/meta.yaml:102]: I'm referring to the discussion in https://github.com/pytorch/pytorch/issues/71280 and particularly [this](https://github.com/pytorch/pytorch/pull/71902#issuecomment-1036250128) comment.

I don't think the vendoring has happened yet, but based on https://github.com/pytorch/pytorch/pull/71902 we'd be able to replace `setuptools` with `packaging`. Unfortunately however, it [seems](https://github.com/pytorch/pytorch/commits/release/1.11) that this was not backported before 1.11 was cut.

Comment (hmaarrfk) [reply recipe/meta.yaml:102]: so lets keep dependijg on it for now?

Comment (h-vetinari) [reply recipe/meta.yaml:102]: Fine by me - 1.11.0 is out as-is. If https://github.com/pytorch/pytorch/pull/71902 ends up getting backported, we can still adapt for 1.11.1.

**Comment by hmaarrfk** [comment]
Attaching logs.
[linxu_64_cuda_logs.zip](https://github.com/conda-forge/pytorch-cpu-feedstock/files/8281108/linxu_64_cuda_logs.zip)

Packages parked at https://anaconda.org/mark.harfouche/ under the label `cfep03`. They will likely stay there for a few days.

Most feedstocks have been migrated to use the pinning, so once this comes out, we should be good to upgrade to 1.11 smoothly?!?!

**Comment by hmaarrfk** [comment]
```
$ anaconda copy --from-label cfep03 --to-label main --to-owner conda-forge mark.harfouche/pytorch/1.11.0
Using Anaconda API: https://api.anaconda.org
Copied file: linux-64/pytorch-1.11.0-cuda102py310h6242262_0.tar.bz2
Copied file: linux-64/pytorch-1.11.0-cuda102py37hb1831c0_0.tar.bz2
Copied file: linux-64/pytorch-1.11.0-cuda102py38hae5ad6b_0.tar.bz2
Copied file: linux-64/pytorch-1.11.0-cuda102py39hfe0cb5b_0.tar.bz2
Copied file: linux-64/pytorch-1.11.0-cuda110py310hac6ddf9_0.tar.bz2
Copied file: linux-64/pytorch-1.11.0-cuda110py37h6337ffc_0.tar.bz2
Copied file: linux-64/pytorch-1.11.0-cuda110py38hade5236_0.tar.bz2
Copied file: linux-64/pytorch-1.11.0-cuda110py39he47eb21_0.tar.bz2
Copied file: linux-64/pytorch-1.11.0-cuda111py310hdfc4b5c_0.tar.bz2
Copied file: linux-64/pytorch-1.11.0-cuda111py37h2eaf6d3_0.tar.bz2
Copied file: linux-64/pytorch-1.11.0-cuda111py38hf27a110_0.tar.bz2
Copied file: linux-64/pytorch-1.11.0-cuda111py39h0b78595_0.tar.bz2
Copied file: linux-64/pytorch-1.11.0-cuda112py310h5c3c7ff_0.tar.bz2
Copied file: linux-64/pytorch-1.11.0-cuda112py37h1c59e6c_0.tar.bz2
Copied file: linux-64/pytorch-1.11.0-cuda112py38h16f0ee0_0.tar.bz2
Copied file: linux-64/pytorch-1.11.0-cuda112py39hda16ee4_0.tar.bz2
Copied 16 files!
```

**Pull request merged**